src/analysis/processing/qgsalgorithmpointstopaths.cpp

```diff
@@ -0,0 +1,334 @@
+  if ( ! textDir.isEmpty() &&
+       ! QDir( textDir ).exists() )
+    throw QgsProcessingException( QObject::tr( "The text output directory does not exist" ) );
+
+  // Store the points in a hash with the group identifier as the key
+  QHash< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > allPoints;
+
+  QgsFeatureRequest request = QgsFeatureRequest().setSubsetOfAttributes( requiredFields, source->fields() );
+  QgsFeatureIterator fit = source->getFeatures( request, QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+  QgsFeature f;
+  const double totalPoints = source->featureCount() > 0 ? 100.0 / source->featureCount() : 0;
+  long currentPoint = 0;
+  feedback->setProgressText( QObject::tr( "Loading points…" ) );
+  while ( fit.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( currentPoint * totalPoints );
+
+    if ( f.hasGeometry() && ! f.geometry().isNull() )
+    {
+      expressionContext.setFeature( f );
+      const QVariant orderValue = orderExpression.evaluate( &expressionContext );
+      const QVariant groupValue = groupExpressionString.isEmpty() ? QVariant() : groupExpression.evaluate( &expressionContext );
+
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
+      const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet()->clone() );
+      allPoints[ groupValue ] << QPair< QVariant, const QgsPoint *>( orderValue, point );
+    }
+    ++currentPoint;
+  }
+
+  int pathCount = 0;
+  currentPoint = 0;
+  QHashIterator< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > hit( allPoints );
+  feedback->setProgressText( QObject::tr( "Creating paths…" ) );
+  while ( hit.hasNext() )
+  {
+    hit.next();
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    auto pairs = hit.value();
+
+    if ( naturalSort )
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 [&collator]( const QPair< const QVariant, const QgsPoint * > &pair1,
+                              const QPair< const QVariant, const QgsPoint * > &pair2 )
+      {
+        return collator.compare( pair1.first.toString(), pair2.first.toString() ) < 0;
+      } );
+    }
+    else
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 []( const QPair< const QVariant, const QgsPoint * > &pair1,
+                     const QPair< const QVariant, const QgsPoint * > &pair2 )
+      {
+        return qgsVariantLessThan( pair1.first, pair2.first );
+      } );
+    }
+
+
+    QVector<QgsPoint> pathPoints;
+    for ( QVector< QPair< QVariant, const QgsPoint * > >::ConstIterator pit = pairs.constBegin(); pit != pairs.constEnd(); ++pit )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+      feedback->setProgress( currentPoint * totalPoints );
```

Comment:
This feedback report should be done up a level, in the loop over the groups (ie. report progress as each output feature is written)

Reply:
I did that on purpose, so that we have a smoother progress bar regardless of the number of groups. If I put it up a level, then a million groups of a few points each will look smooth, but a few groups with a million points each will not!

Reply:
Hm -- i guess ideally we'd do it on both levels -- so we'd report the feedback on the group, and then on the points within the group. If you want to try this QgsProcessingMultiStepFeedback can be used, which allows you to specify the total number of steps and then the individual progress reports are scaled accordingly.

Reply:
If groups have uneven number of points then the progress bar will not have constant speed (think windows saying your copy will take 3 minutes but stays at 99% for 2 hours!) I think that reporting on a per point basis guarantees that the progress bar shows the correct fraction of the total time the process will take. Again, this only matters when points are by the millions, otherwise the progress bar is shown for too little to matter!

Reply:
Ah, I read the code wrong -- you're approach is good! It just needs scaling so it only updates the final 50% of the progress instead of restarting at 0

Reply:
I did that when I changed the progress bar on the first iteration as per your suggestion!